### PR TITLE
Reproducibility 3/3

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Last PR of the series after https://github.com/huggingface/diffusers/pull/1906. 

In this PR, I'll update all tests to use CPU generators for better determinism and also add a reproducible doc page.

### CODE REVIEW

File: docs/source/en/conceptual/reproducibility.mdx [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+In the following, we should how to best control sources of randomness for diffusion models.
+
+## Inference
+
+During inference, diffusion pipelines heavily rely on random sampling operations, such as the creating the 
+gaussian noise tensors to be denoised as well as adding noise to the scheduling step.
+
+Let's have a look at an example. We run the [DDIM pipeline](./api/pipelines/ddim.mdx) 
+for just two inference steps and return a numpy tensor to look into the numerical values of the output.
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:32]:
```suggestion
## Inference

During inference, diffusion pipelines heavily rely on random sampling operations, such as the creating the 
gaussian noise tensors to be denoised and adding noise to the scheduling step.

Let's have a look at an example. We run the [DDIM pipeline](./api/pipelines/ddim.mdx) 
for just two inference steps and return a numpy tensor to look into the numerical values of the output.
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+# run pipeline for just two steps and return numpy tensor
+image = ddim(num_inference_steps=2, output_type="np").images
+print(np.abs(image).sum())
+```
+
+Running the above prints a value of 1464.2076, but running it again prints a different 
+value of 1495.1768. What is going on here? Every time the pipeline is run, gaussian noise 
+is created and step-wise denoised. To create the gaussian noise with [`torch.randn`](https://pytorch.org/docs/stable/generated/torch.randn.html) a different random seed is taken every time, thus leading to a different result.
+This is a desired property of diffusion pipelines as it means that the pipeline can create a different random image every time it is run. In many cases, one would like to generate the exact same image of a certain 
+run, for which case an instance of a [PyTorch generator](https://pytorch.org/docs/stable/generated/torch.randn.html) has to be passed:
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:52]:
```suggestion
Running the above prints a value of 1464.2076, but running it again prints a different 
value of 1495.1768. What is going on here? Every time the pipeline is run, gaussian noise 
is created and step-wise denoised. To create the gaussian noise with [`torch.randn`](https://pytorch.org/docs/stable/generated/torch.randn.html), a different random seed is taken every time, thus leading to a different result.
This is a desired property of diffusion pipelines, as it means that the pipeline can create a different random image every time it is run. In many cases, one would like to generate the exact same image of a certain 
run, for which case an instance of a [PyTorch generator](https://pytorch.org/docs/stable/generated/torch.randn.html) has to be passed:
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+# run pipeline for just two steps and return numpy tensor
+image = ddim(num_inference_steps=2, output_type="np", generator=generator).images
+print(np.abs(image).sum())
+```
+
+Running the above always prints a value of 1491.1711 - also upon running it again because we 
+define the generator object to be passed to all random functions of the pipeline.
+
+If you run this code snippet on your specific hardware and version, you should get a very similar, if not the same result.
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:75]:
```suggestion
Running the above always prints a value of 1491.1711 - also upon running it again because we 
define the generator object to be passed to all random functions of the pipeline.

If you run this code snippet on your specific hardware and version, you should get a similar, if not the same, result.
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+</Tip>
+
+Great! Now, we know how to write reproducible pipelines, but it does get a bit tricker since the above example only runs on CPU. How do we also achieve reproducibility on GPU? 
+In short, one should not expect to get full reproducibility across hardwares when running pipelines on GPU 
+as matrix multiplications are less deterministic on GPU than on CPU and diffusion pipelines tend to require
+a lot of matrix multiplications. Let's see what we can do though to keep the randomness within limits across 
+different GPU hardware.
+
+To achieve maximum speed performance, it is recommended to create the generator directly on GPU when running 
+the pipeline on GPU:
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:93]:
```suggestion
<Tip>

It might be a bit unintuitive at first to pass `generator` objects to the pipelines instead of 
just integer values representing the seed, but this is the recommended design when dealing with 
probabilistic models in PyTorch as generators are *random states* that are advanced and can thus be 
passed to multiple pipelines in a sequence.

</Tip>

Great! Now, we know how to write reproducible pipelines, but it gets a bit trickier since the above example only runs on the CPU. How do we also achieve reproducibility on GPU? 
In short, one should not expect full reproducibility across different hardware when running pipelines on GPU 
as matrix multiplications are less deterministic on GPU than on CPU and diffusion pipelines tend to require
a lot of matrix multiplications. Let's see what we can do to keep the randomness within limits across 
different GPU hardware.

To achieve maximum speed performance, it is recommended to create the generator directly on GPU when running 
the pipeline on GPU:
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+image = ddim(num_inference_steps=2, output_type="np", generator=generator).images
+print(np.abs(image).sum())
+```
+
+Running the above now prints a value of 1389.8634 - even though we're using the exact same seed! 
+This is unfortunate as it means we cannot really reproduce results we achieved on GPU, also on CPU.
+Nevertheless, it should be expected since the GPU uses a different random number generator than the CPU.
+
+To circumvent this problem, we created a [~`randn_tensor`] function which can create random noise 
+on CPU and then move the tensor to GPU if necessary. The function is used every where inside the pipelines allowing the user to **always** pass a CPU generator even if the pipeline is run on GPU:
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:119]:
```suggestion
Running the above now prints a value of 1389.8634 - even though we're using the exact same seed! 
This is unfortunate as it means we cannot reproduce the results we achieved on GPU, also on CPU.
Nevertheless, it should be expected since the GPU uses a different random number generator than the CPU.

To circumvent this problem, we created a [~`randn_tensor`] function, which can create random noise 
on the CPU and then move the tensor to GPU if necessary. The function is used everywhere inside the pipelines allowing the user to **always** pass a CPU generator even if the pipeline is run on GPU:
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+As a consequence, we recommend to always pass a CPU generator if Reproducibility is important.
+The loss of performance is often neglectable, but one can be sure to generate much more similar 
+values than if the pipeline would have been run on CPU.
+
+</Tip>
+
+Finally, we noticed that more complex pipelines, such as [`UnCLIPPipeline`] are often extremely 
+susceptible precision error propagation and thus one cannot expect even similar results across 
+different GPU hardware or PyTorch versions. In such cases, one simply has to make sure to run 
+exactly the same hardware and PyTorch version for full Reproducibility.
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:154]:
```suggestion
Running the above now prints a value of 1491.1713, much closer to the value of 1491.1711 when 
the pipeline is fully run on the CPU.

<Tip>

As a consequence, we recommend always passing a CPU generator if Reproducibility is important.
The loss of performance is often neglectable, but one can be sure to generate much more similar 
values than if the pipeline would have been run on CPU.

</Tip>

Finally, we noticed that more complex pipelines, such as [`UnCLIPPipeline`] are often extremely 
susceptible to precision error propagation and thus one cannot expect even similar results across 
different GPU hardware or PyTorch versions. In such cases, one has to make sure to run 
exactly the same hardware and PyTorch version for full Reproducibility.
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+## Randomness utilities
+
+### randn_tensor
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:158]: Very cool doc!

File: src/diffusers/utils/torch_utils.py [hunk 8]
```diff
@@ -65,6 +65,7 @@ def randn_tensor(
         ]
         latents = torch.cat(latents, dim=0).to(device)
     else:
+        print(rand_device)
```
Comment (patil-suraj) [inline src/diffusers/utils/torch_utils.py:68]:
```suggestion
```

File: test_corrections.txt [hunk 9]
```diff
@@ -0,0 +1,3 @@
+tests/pipelines/stable_diffusion/test_stable_diffusion_img2img.py;StableDiffusionImg2ImgPipelineSlowTests;test_stable_diffusion_img2img_pipeline_multiple_of_8;expected_slice = np.array([0.9393, 0.9500, 0.9399, 0.9438, 0.9458, 0.9400, 0.9455, 0.9414, 0.9423])
+tests/pipelines/stable_diffusion/test_stable_diffusion_instruction_pix2pix.py;StableDiffusionInstructPix2PixPipelineSlowTests;test_stable_diffusion_pix2pix_pipeline_multiple_of_8;expected_slice = np.array([0.2726, 0.2529, 0.2664, 0.2655, 0.2641, 0.2642, 0.2591, 0.2649, 0.2590])
+tests/pipelines/stable_diffusion/test_stable_diffusion_k_diffusion.py;StableDiffusionPipelineIntegrationTests;test_stable_diffusion_2;expected_slice = np.array([0.1237, 0.1320, 0.1438, 0.1359, 0.1390, 0.1132, 0.1277, 0.1175, 0.1112])
```
Comment (patil-suraj) [inline test_corrections.txt:3]: probably leftover from debugging, and should be deleted.

File: utils/overwrite_expected_slice.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team.
```
Comment (patil-suraj) [inline utils/overwrite_expected_slice.py:2]:
```suggestion
# Copyright 2023 The HuggingFace Inc. team.
```

File: docs/source/en/conceptual/reproducibility.mdx [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+import numpy as np
+
+model_id = "google/ddpm-cifar10-32"
+
+# load model and scheduler
+ddim = DDIMPipeline.from_pretrained(model_id)
+ddim.to("cuda")
+
+# create a generator for reproducibility
+generator = torch.manual_seed(0)
```
Comment (patil-suraj) [inline docs/source/en/conceptual/reproducibility.mdx:133]: Should we still use a `torch.Generator` here? Using `manual_seed` will set the global seed, no?

**Review by patil-suraj — APPROVED** [review]
Great PR! It must have taken a lot of time to change all these tests! I mostly left some nits.

File: docs/source/en/conceptual/reproducibility.mdx (hunk 11 above)
Comment (patrickvonplaten) [reply docs/source/en/conceptual/reproducibility.mdx:133]: Think `torch.Generator` does that as well

File: docs/source/en/conceptual/reproducibility.mdx [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+from diffusers import DDIMPipeline
+import numpy as np
+
+model_id = "google/ddpm-cifar10-32"
+
+# load model and scheduler
+ddim = DDIMPipeline.from_pretrained(model_id)
+
+# create a generator for reproducibility
+generator = torch.Generator(device="cpu").manual_seed(0)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply docs/source/en/conceptual/reproducibility.mdx:65]: Yeah but maybe a bit more readability this way, but both works I guess

File: src/diffusers/utils/testing_utils.py [hunk 13]
```diff
@@ -45,6 +45,21 @@ def torch_all_close(a, b, *args, **kwargs):
     return True
 
 
+def print_tensor_test(tensor, filename="test_corrections.txt", expected_tensor_name="expected_slice"):
```
Comment (patrickvonplaten) [inline src/diffusers/utils/testing_utils.py:48]: BTW @pcuenca @patil-suraj @yiyixuxu @williamberman 

I added some functionality that helps to automatically correct slice-comparison tests. 
All you need to do here is to add:

```py
print_tensor_test(image_slice)
```

and it'll write the correct tensor to a file: `filename` with the name `expected_tensor_name`.

The utils file here: https://github.com/huggingface/diffusers/pull/1924/files#diff-bd9ba404f47129f2218fc94c5e528cc81bfd10cef1eff1d6d543f6ffa0143367 can then take this file and will automatically overwrite the test with the correct results. This should save quite some time going forward :-)

Comment (patil-suraj) [reply src/diffusers/utils/testing_utils.py:48]: That's very cool, thanks a lot !
